File: Sources/ATTNCreativeUrlFormatter.m
//
//  CreativeUrlFormatter.m
//  attentive-ios-sdk
//
//  Created by Olivia Kim on 3/9/23.
//

#import "ATTNCreativeUrlFormatter.h"
#import "Internal/ATTNAppInfo.h"

@implementation ATTNCreativeUrlFormatter

+ (nonnull NSString *)buildCompanyCreativeUrlForDomain:(NSString *)domain
                                                  mode:(NSString *)mode
                                          userIdentity:(ATTNUserIdentity *)userIdentity {

  NSURLComponents *components = [[NSURLComponents alloc] initWithString:@"https://creatives.attn.tv/mobile-apps/index.html"];

  // Add query parameters
  NSMutableArray *queryItems = [[NSMutableArray alloc] initWithObjects:
                                                           [[NSURLQueryItem alloc] initWithName:@"domain"
                                                                                          value:domain],
                                                           nil];

  if ([mode isEqualToString:@"debug"]) {
    [queryItems addObject:[[NSURLQueryItem alloc] initWithName:mode value:@"matter-trip-grass-symbol"]];
  }

  if (userIdentity.visitorId != nil) {
    [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"vid" value:userIdentity.visitorId]];
  } else {
    NSLog(@"ERROR: No Visitor ID Found. This should not happen.");
  }

  if (userIdentity.identifiers[IDENTIFIER_TYPE_CLIENT_USER_ID] != nil) {
    [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"cuid" value:userIdentity.identifiers[IDENTIFIER_TYPE_CLIENT_USER_ID]]];
  }

  if (userIdentity.identifiers[IDENTIFIER_TYPE_PHONE] != nil) {
    [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"p" value:userIdentity.identifiers[IDENTIFIER_TYPE_PHONE]]];
  }

  if (userIdentity.identifiers[IDENTIFIER_TYPE_EMAIL] != nil) {
    [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"e" value:userIdentity.identifiers[IDENTIFIER_TYPE_EMAIL]]];
  }

  if (userIdentity.identifiers[IDENTIFIER_TYPE_KLAVIYO_ID] != nil) {
    [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"kid" value:userIdentity.identifiers[IDENTIFIER_TYPE_KLAVIYO_ID]]];
  }

  if (userIdentity.identifiers[IDENTIFIER_TYPE_SHOPIFY_ID] != nil) {
    [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"sid" value:userIdentity.identifiers[IDENTIFIER_TYPE_KLAVIYO_ID]]];
  }

  if (userIdentity.identifiers[IDENTIFIER_TYPE_CUSTOM_IDENTIFIERS] != nil) {
    [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"cstm" value:[[self class] getCustomIdentifiersJson:userIdentity]]];
  }

  // Add SDK info just for analytics purposes
  [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"sdkVersion" value:[ATTNAppInfo getSdkVersion]]];
  [queryItems addObject:[[NSURLQueryItem alloc] initWithName:@"sdkName" value:[ATTNAppInfo getSdkName]]];

  [components setQueryItems:queryItems];

  return [components string];
}

+ (nonnull NSString *)getCustomIdentifiersJson:(ATTNUserIdentity *)userIdentity {
  @try {
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:userIdentity.identifiers[IDENTIFIER_TYPE_CUSTOM_IDENTIFIERS]
                                                       options:0 // Do not pretty print the json string
                                                         error:&error];
    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
  } @catch (NSException *exception) {
    NSLog(@"ERROR: Could not parse custom identifiers to json %@", exception);
  }

  return @"{}";
}

@end

File: Sources/Migration/ATTNCreativeUrlFormatter.swift
//
//  ATTNCreativeUrlFormatter.swift
//  attentive-ios-sdk-framework
//
//  Created by Vladimir - Work on 2024-05-29.
//

import Foundation

@objc(ATTNCreativeUrlFormatter)
public final class ATTNCreativeUrlFormatter: NSObject {
  private enum Constants {
    static var scheme: String { "https" }
    static var host: String { "creatives.attn.tv" }
    static var path: String { "/mobile-apps/index.html" }
  }

  @objc(buildCompanyCreativeUrlForDomain:mode:userIdentity:)
  public static func buildCompanyCreativeUrl(
    forDomain domain: String,
    mode: String,
    userIdentity: ATTNUserIdentity
  ) -> String {
    var components = URLComponents()
    components.scheme = Constants.scheme
    components.host = Constants.host
    components.path = Constants.path

    var queryItems = [
      URLQueryItem(name: "domain", value: domain)
    ]

    if mode == "debug" {
      queryItems.append(URLQueryItem(name: mode, value: "matter-trip-grass-symbol"))
    }

    queryItems.append(URLQueryItem(name: "vid", value: userIdentity.visitorId))

    if let clientUserId = userIdentity.identifiers[IDENTIFIER_TYPE_CLIENT_USER_ID] as? String {
      queryItems.append(URLQueryItem(name: "cuid", value: clientUserId))
    }

    if let phone = userIdentity.identifiers[IDENTIFIER_TYPE_PHONE] as? String {
      queryItems.append(URLQueryItem(name: "p", value: phone))
    }

    if let email = userIdentity.identifiers[IDENTIFIER_TYPE_EMAIL] as? String {
      queryItems.append(URLQueryItem(name: "e", value: email))
    }

    if let klaviyoId = userIdentity.identifiers[IDENTIFIER_TYPE_KLAVIYO_ID] as? String {
      queryItems.append(URLQueryItem(name: "kid", value: klaviyoId))
    }

    if let shopifyId = userIdentity.identifiers[IDENTIFIER_TYPE_SHOPIFY_ID] as? String {
      queryItems.append(URLQueryItem(name: "sid", value: shopifyId))
    }

    if let customIdentifiersJson = getCustomIdentifiersJson(userIdentity: userIdentity) {
      queryItems.append(URLQueryItem(name: "cstm", value: customIdentifiersJson))
    }

    // Add SDK info just for analytics purposes
    queryItems.append(URLQueryItem(name: "sdkVersion", value: ATTNAppInfo.getSdkVersion()))
    queryItems.append(URLQueryItem(name: "sdkName", value: ATTNAppInfo.getSdkName()))

    components.queryItems = queryItems

    return components.string ?? ""
  }
}

fileprivate extension ATTNCreativeUrlFormatter {
  static func getCustomIdentifiersJson(userIdentity: ATTNUserIdentity) -> String? {
    do {
      guard let customIdentifiers = userIdentity.identifiers[IDENTIFIER_TYPE_CUSTOM_IDENTIFIERS] else { return nil }
      let jsonData = try JSONSerialization.data(withJSONObject: customIdentifiers, options: [])
      return String(data: jsonData, encoding: .utf8)
    } catch {
      print("ERROR: Could not parse custom identifiers to json \(error)")
    }
    return nil
  }
}
